core/storage/trie/serialization/trie_serializer_impl.cpp

```diff
@@ -55,39 +56,44 @@ namespace kagome::storage::trie {
   outcome::result<RootHash> TrieSerializerImpl::storeRootNode(
       TrieNode &node, StateVersion version) {
     auto batch = backend_->batch();
+    BOOST_ASSERT(batch != nullptr);
 
-    OUTCOME_TRY(enc,
-                codec_->encodeNode(node,
-                                   version,
-                                   [&](const TrieNode *,
-                                       common::BufferView hash,
-                                       common::Buffer &&encoded) {
-                                     return batch->put(hash,
-                                                       std::move(encoded));
-                                   }));
-    auto key = codec_->hash256(enc);
-    OUTCOME_TRY(batch->put(key, std::move(enc)));
+    OUTCOME_TRY(
+        enc,
+        codec_->encodeNode(node,
+                           version,
+                           [&](TrieNode const &node,
+                               common::BufferView hash,
+                               common::Buffer &&encoded) {
+                             current_stats_.new_nodes_written++;
+                             if (node.getValue()) {
+                               current_stats_.values_written++;
+                             }
+                             return batch->put(hash, std::move(encoded));
+                           }));
+    auto hash = codec_->hash256(enc);
+    OUTCOME_TRY(batch->put(hash, std::move(enc)));
     OUTCOME_TRY(batch->commit());
 
-    return key;
+    return hash;
   }
 
   outcome::result<PolkadotTrie::NodePtr> TrieSerializerImpl::retrieveNode(
-      const std::shared_ptr<OpaqueTrieNode> &parent,
+      const std::shared_ptr<OpaqueTrieNode> &node,
       const OnNodeLoaded &on_node_loaded) const {
-    if (auto p = std::dynamic_pointer_cast<DummyValue>(parent); p != nullptr) {
+    if (auto p = std::dynamic_pointer_cast<DummyValue>(node); p != nullptr) {
       OUTCOME_TRY(value, backend_->get(*p->value.hash));
       if (on_node_loaded) {
         on_node_loaded(value);
       }
       p->value.value = value.into();
       return nullptr;
     }
-    if (auto p = std::dynamic_pointer_cast<DummyNode>(parent); p != nullptr) {
+    if (auto p = std::dynamic_pointer_cast<DummyNode>(node); p != nullptr) {
```

Comment:
Is this condition right? It continues the previous one

Reply:
No, previous was `DummyValue`, not `DummyNode`